File: src/lib/config_manager.py
"""Configuration manager for Zephyr Desktop.

Manages the ~/.zephyr/ configuration directory and provides atomic
JSON file persistence. The config dir is injectable for testability.
"""

import json
import logging
import os
import tempfile
from pathlib import Path

logger = logging.getLogger(__name__)

DEFAULT_CONFIG_DIR = Path.home() / ".zephyr"


class ConfigManager:
    """Manages the Zephyr configuration directory and JSON file I/O.

    Args:
        config_dir: Path to configuration directory. Defaults to ~/.zephyr/.
    """

    def __init__(self, config_dir: Path | None = None):
        self._config_dir = config_dir or DEFAULT_CONFIG_DIR

    def get_config_dir(self) -> Path:
        """Return the configuration directory path."""
        return self._config_dir

    def ensure_config_dir(self) -> Path:
        """Create the configuration directory if it doesn't exist.

        Returns:
            Path to the configuration directory.
        """
        self._config_dir.mkdir(parents=True, exist_ok=True)
        return self._config_dir

    def load_json(self, filename: str) -> dict:
        """Load a JSON file from the configuration directory.

        Args:
            filename: Name of the JSON file (e.g. 'projects.json').

        Returns:
            Parsed dict, or empty dict if file is missing or invalid.
        """
        filepath = self._config_dir / filename
        if not filepath.exists():
            return {}
        try:
            with open(filepath, "r", encoding="utf-8") as f:
                data = json.load(f)
            if not isinstance(data, dict):
                logger.warning(
                    "Expected dict in %s, got %s; returning empty dict",
                    filepath,
                    type(data).__name__,
                )
                return {}
            return data
        except (json.JSONDecodeError, OSError) as exc:
            logger.warning("Failed to load %s: %s", filepath, exc)
            return {}

    def save_json(self, filename: str, data: dict) -> None:
        """Atomically write a dict as JSON to the configuration directory.

        Uses a temporary file + rename to prevent corruption from partial
        writes (e.g. if the process is killed mid-write).

        Args:
            filename: Name of the JSON file (e.g. 'projects.json').
            data: Dictionary to serialize.
        """
        self.ensure_config_dir()
        filepath = self._config_dir / filename

        # Write to a temp file in the same directory, then atomically rename.
        # dir= ensures the temp file is on the same filesystem as the target.
        fd, tmp_path = tempfile.mkstemp(
            dir=self._config_dir, suffix=".tmp", prefix=f".{filename}."
        )
        try:
            with os.fdopen(fd, "w", encoding="utf-8") as f:
                json.dump(data, f, indent=2, ensure_ascii=False)
                f.flush()
                os.fsync(f.fileno())
            os.replace(tmp_path, filepath)
        except BaseException:
            # Clean up the temp file on any failure.
            try:
                os.unlink(tmp_path)
            except OSError:
                pass
            raise

File: src/services/config-manager.ts
/**
 * ConfigManager — manages ~/.zephyr/ config directory and JSON file I/O.
 *
 * Runs in the Electron main process. The config directory is injectable
 * for testability — tests pass a temp dir instead of the real ~/.zephyr/.
 *
 * Atomic writes prevent data corruption: data is written to a .tmp file
 * first, then renamed over the destination atomically.
 */

import fs from 'fs';
import os from 'os';
import path from 'path';

const DEFAULT_CONFIG_DIR = path.join(os.homedir(), '.zephyr');

export class ConfigManager {
  private readonly configDir: string;

  constructor(configDir?: string) {
    this.configDir = configDir ?? DEFAULT_CONFIG_DIR;
  }

  /** Returns the config directory path. */
  getConfigDir(): string {
    return this.configDir;
  }

  /** Creates the config directory if it doesn't exist. */
  ensureConfigDir(): void {
    fs.mkdirSync(this.configDir, { recursive: true });
  }

  /**
   * Reads and parses a JSON file from the config directory.
   * Returns null if the file does not exist or cannot be parsed.
   */
  loadJson<T>(filename: string): T | null {
    const filepath = path.join(this.configDir, filename);
    try {
      const raw = fs.readFileSync(filepath, 'utf-8');
      return JSON.parse(raw) as T;
    } catch (err: unknown) {
      // File missing (ENOENT) or unreadable → return null (expected case)
      // Corrupt JSON → return null (safe degradation)
      if (isNodeError(err) && err.code !== 'ENOENT') {
        // Log unexpected errors but still degrade gracefully
        // eslint-disable-next-line no-console
        console.warn(`[ConfigManager] Failed to load ${filepath}:`, err);
      }
      return null;
    }
  }

  /**
   * Atomically writes data as JSON to the config directory.
   *
   * The write goes to a .tmp file first, then fs.renameSync replaces the
   * target file. This guarantees the reader always sees a complete file,
   * even if the process is killed mid-write.
   */
  saveJson(filename: string, data: unknown): void {
    this.ensureConfigDir();

    const filepath = path.join(this.configDir, filename);
    const tmpPath = filepath + '.tmp';

    try {
      fs.writeFileSync(tmpPath, JSON.stringify(data, null, 2), 'utf-8');
      fs.renameSync(tmpPath, filepath);
    } catch (err) {
      // Clean up temp file on failure
      try {
        fs.unlinkSync(tmpPath);
      } catch {
        // Ignore cleanup errors
      }
      throw err;
    }
  }
}

function isNodeError(err: unknown): err is NodeJS.ErrnoException {
  return err instanceof Error && 'code' in err;
}
